woocommerce/woocommerce-ios
# Dashboard: fetch stats data only for the visible time range tab

The main part of #6865 


Sorry about the 500+ diffs 😅 the deleted code is mostly removing unused code and moving code! I tried explaining the changes in the PR description, but please lemme know if anything is confusing or unclear.

### Description


#### Current dashboard behavior

When we implemented the dashboard with the new stats v4 endpoints (`wc-analytics`) in 2019, we had to support merchants who haven't upgraded their WC plugin to the version with the stats v4 support. To check if the site supports stats v4, we make an API request to [fetch the yearly stats and also persist the latest stats version in app settings](https://github.com/woocommerce/woocommerce-ios/blob/91acf263d2ffbbc52c7e36abc4c0e307df86b04d/WooCommerce/Classes/ViewRelated/Dashboard/Factories/StatsVersionCoordinator.swift#L30-L53). Also, we sync stats data for all 4 time range tabs on the following timing:
- App launch
- Pull-to-refresh on any time range tab
- Switching from another main tab to the My store tab (in `viewWillAppear`)

#### Why we want to change the current behavior

Back in 2019, the app had way fewer features than today and we didn't have the performance issue where so many API requests are fired on app launch and later on. All these API requests make the app feel slow, and the dashboard contributes to a bunch of them on app launch - at least 12 (3 requests * 4 time range tabs) plus 1 availability check. A bunch of merchants mentioned the performance issue, and our API team also noticed the higher stats requests in iOS than Android.

#### Android behavior

To match the behavior in Android, we're changing the behavior to the following:
- App launch: only fetch data for the visible time range tab (Today tab for now, maybe we will persist and show the latest time range tab in the future)
- Pull-to-refresh on any time range tab: only fetch data for the visible time range tab
- Switching from another main tab to the My store tab: we don't fetch data anymore
- Switching time range tab: only fetch data for the selected time range tab

In order to check if stats v4 data are available, Android just [fetches the stats data normally and show the "no stats" UI if the store stats API request returns an error that indicates the data are unavailable](https://github.com/woocommerce/woocommerce-android/blob/e602fe1d0569b8200d91015ca225d8624af4b5fb/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreFragment.kt#L179-L186).

#### Major changes

Because of this change, a lot of the stats v4 availability check code can be deleted (there's more to delete but I thought the PR is already large enough). The major high-level change is to change the current behavior:

app launch/pull-to-refresh/switching tabs --> yearly store stats API request
--> if success: show `StoreStatsAndTopPerformersViewController` and sync data with store stats / top performers / visitor stats API requests for all 4 tabs
--> otherwise: show `DeprecatedDashboardStatsViewController`

to the new behavior:

app launch/pull-to-refresh --> show `StoreStatsAndTopPerformersViewController` by default --> make store stats / top performers / visitor stats API requests for the visible tab
--> if store stats API request returns `rest_no_route` error: show `DeprecatedDashboardStatsViewController`
--> otherwise: show `StoreStatsAndTopPerformersViewController`

#### Pull-to-refresh support in `DeprecatedDashboardStatsViewController`

Because we no longer fetch stats data when switching tabs (in `viewWillAppear`), when `DeprecatedDashboardStatsViewController` is shown there is currently no way to refresh data. To support pull-to-refresh in `DeprecatedDashboardStatsViewController`, I added a scroll view with a refresh control to encapsulate the empty state child view controller. In order to know when to end animating the refresh control, I added `async/await` to the pull-to-refresh and data reload code. I'm still new to Swift concurrency, so please feel free to suggest any changes 😄 . The only caveat I found is not to add `async` to an `@objc` function, otherwise it crashes when it's triggered.

#### Detecting when the user selects a time range tab

We're using `XLPagerTabStrip` library for the tabbed dashboard UI, and the only way I found is to implement `updateIndicator(for:fromIndex:toIndex:withProgressPercentage:indexWasChanged:)` in `PagerTabStripIsProgressiveDelegate` which `ButtonBarPagerTabStripViewController` (`StoreStatsAndTopPerformersViewController`'s base class) implements. We have plans to repace this third-party library in p91TBi-47N-p2 https://github.com/woocommerce/woocommerce-ios/issues/1911, but it's a big undertaking and we haven't got to make this switch.

#### Stats availability changes 

The previous availably checks are mostly in `StatsVersionCoordinator` and `DashboardUIFactory`. In order not to make too many changes while maintaining testability for the new data fetching behavior, I created a new class `DashboardViewModel` which just contains 3 data fetching functions for store stats / top performers / visitor stats because the logic around it is too coupled with UI in `StoreStatsAndTopPerformersViewController`. I'm open to any suggestions here as well 🙂 

### Testing instructions


It's the easiest to use a proxy tool (e.g. Charles) to see the API requests fired from the following actions. The app should only fetch stats (`wc-analytics/leaderboards`, `stats/visits`, `wc-analytics/reports/revenue/stats`) once per type for the currently visible time range tab whenever it's refreshing data.

- Log in to a store if needed --> only 3 stats API requests are made for the "Today" tab
- Tapping on a different time range tab --> only 3 stats API requests are made for the selected tab
- Pulling down to refresh --> only 3 stats API requests are made for the "Today" tab
- In `wp-admin`, go to WooCommerce > Settings > Advanced and disable WC Analytics
- In the app, pull to refresh the visible tab --> only 3 stats API requests are made for the selected tab, and then the UI changes to the deprecated stats UI
- In `wp-admin`, re-enable WC Analytics
- In the app, pull to refresh the visible tab --> only 3 stats API requests are made for the selected tab, and then the UI changes back to the stats UI

---
